### DISCUSSION TITLE
[Feature Request] Filter store by type of item

### DISCUSSION THREAD

**Discussion by superm1991** [ChristopherKlay/StadiaEnhanced#49] [Ideas & Requests]
Hi!

Can you add filters in the store so that we can choose to hide items of type Game, or of type Bundle, or of type Add-on, or combinations of those?

It looks like this info is clearly available in the webpage so it shouldn't be too difficult:
```
<div class="vaa0f eoQBOd">Game</div>
<div class="vaa0f eoQBOd">Bundle</div>
<div class="vaa0f eoQBOd">Add-on</div>
```

Thanks!

**Comment by ChristopherKlay** [comment]
I already thought about this, but it comes with a couple of issues.

1. The "All Games" list used for searching doesn't actually include add-ons, so it would return a empty list if you only search for those.
2. There isn't actually a complete list of all games/bundles/add-ons available that can be filtered.
3. While games are easily filtered, categories in the store don't actually include such tags.
4. Games are labeled as "bundle" and such, but said label is based on the currently used language.
5. 
I'll definitely look into it when i'm done with the new settings menu after the holidays however. I wanted to add some store/search related stuff anyway, like a quick access to the newly tagged categories (:

**Reply by superm1991** [reply]
Thanks for your answer! I hadn't considered all of that.

I'd be happy with something as simple as this for a first drop:
- Filter is only active if URL is equal to "https://stadia.google.com/store/list/14" (the "On Sale" page)
- Filter feature is only available if language is English
- Only 1 filter: hide Bundle/Add-on, show only Standard Editions of games

But it's okay if you have other priorities right now :)

Good luck with the settings menu, and happy holidays!

**Comment by superm1991** [comment]
Hi Christopher,

I hope your holidays are going well.

I've tested some code, inspired by what you did for the search bar:
```
    // Store Search - Type of item (by superm1991)
    if (document.location.href.indexOf("/store/list/3?") != -1 || document.location.href.indexOf("/store/list/14?") != -1) {
        var enhanced_SearchTypeTerm = new URLSearchParams(window.location.search).get('searchType');
        if (enhanced_SearchTypeTerm !== null) {
            var enhanced_SearchTypeResultCount = 0;
            document.querySelector(".TJQxhf").style.display = "none";
            document.getElementsByClassName("h6J22d null QAAyWd");
            enhanced_SearchTypeResults = document.getElementsByClassName("h6J22d");
            for (var i = 0; i < enhanced_SearchTypeResults.length; i++) {
                enhanced_TypeOfItem = enhanced_SearchTypeResults[i].getElementsByClassName("vaa0f eoQBOd");
                if (enhanced_TypeOfItem.length > 0 && enhanced_TypeOfItem[0].textContent.match(new RegExp(enhanced_SearchTypeTerm, 'gi')) === null) {
                    enhanced_SearchTypeResults[i].style.display = "none";
                } else {
                    enhanced_SearchTypeResultCount += 1;
                }
            }
            document.querySelector(".HZ5mJ").innerHTML = "Store items of type '" + enhanced_SearchTypeTerm + "' (" + enhanced_SearchTypeResultCount + ")";
        }
    }
```

It allows the user to make URLs with '?searchType=...' (whatever the type is in their language) and filter the store items.

Do you think you can add this to your extension?
The UI can wait, but at least we would have the feature!

**Comment by ChristopherKlay** [comment]
The issue here is that `/store/list/3` only lists "Game" and "Bundle" types in the first place, add-ons aren't on the list to begin with, so you can't really filter out anything but "bundles", which also filters out some base games as well, because for example "Hitman 2" is listed as a "Bundle", due to being packed with the "World of Assassination" hub to start the title.

I'm currently finishing the menu changes (see; #53) and will implement a couple of store related features hopefully including a filter for bundles/add-ons and such afterwards.

I really just want to finish one point first, before starting to work on a different aspect of the extension, so i don't end up having to patch up "quick solutions" later on.

**Reply by superm1991** [reply]
Perfectly understandable!
Sorry for pushing, I just wanted to help :)

> you can't really filter out anything but "bundles"

That's exactly what I'd want on the "All Games" page :)

> which also filters out some base games as well

Yeah... To fix this, I would make a list/hashtable of Games + a list/hashtable of Bundles, and hide only Bundles that do not appear in the list of Games. But that's kind of a heavy dev...

I'll let you work at your own pace and on your own schedule. Main menu ([#53](https://github.com/ChristopherKlay/StadiaEnhanced/discussions/53)) is looking good! :)

(PS: I understand if you'd prefer to work alone, but if that's not the case... don't hesitate to let me know if you want some help on anything or if you want to brainstorm some ideas.)

**Comment by ChristopherKlay** [comment]
Looking into the possibility to filter out bundles and other aspects right now, but it doesn't look like it's currently possible, since the only way to filter out items is based on text content which in return is different based on translations.

I really don't want to restrict a function to a given language and there currently isn't any way to filter out specific categories based on their markup outside of that. The only reliable option here would be to filter items against a pre-loaded array of ID's (which are present in the store entries) and hide bundles based on that - but that would be a tremendous amount of work for a rather small feature, that still breaks whenever the list isn't updated on new releases.

I'm definitely still looking for a solution on this, but i won't implement this until it's functionality can be provided by data that isn't purely translation based, or requires stored data to be checked correctly.

**Reply by superm1991** [reply]
Hi Christopher,

I think I found the solution! :)

All items of type Game seem to have "17:1" in their 'jslog' property.
All items of type Bundle seem to have "17:3" in their 'jslog' property.
All items of type Add-on seem to have "17:2" in their 'jslog' property.

Let me know what you think!!! :)

**Reply by superm1991** [reply]
It looks like this code works as expected:
```
    // Store Search - Type of item
    // Example: /store/list/3?searchType=1
    if (document.location.href.indexOf("/store/list/3?") != -1 || document.location.href.indexOf("/store/list/14?") != -1) {
        var enhanced_SearchTerm = new URLSearchParams(window.location.search).get('searchType');
        if (enhanced_SearchTerm !== null) {
            var enhanced_SearchResultCount = 0;
            document.getElementsByClassName("h6J22d null QAAyWd");
            enhanced_SearchResults = document.getElementsByClassName("h6J22d");
            for (var i = 0; i < enhanced_SearchResults.length; i++) {
                if (enhanced_SearchResults[i].getAttribute("jslog").match(new RegExp("17:" + enhanced_SearchTerm, 'gi')) === null) {
                    enhanced_SearchResults[i].style.display = "none";
                } else {
                    enhanced_SearchResultCount += 1;
                }
            }
            document.querySelector(".HZ5mJ").innerHTML = enhanced_lang.searchheader + " '" + enhanced_SearchTerm + "' (" + enhanced_SearchResultCount + ")";
        }
    }
```

Url "https://stadia.google.com/store/list/3?searchType=1" gives only the Game items (133 found).
Url "https://stadia.google.com/store/list/3?searchType=3" gives only the Bundle items (68 found).

**Comment by ChristopherKlay** [comment]
I'll work on a new patch until tomorrow and will make sure to include this (including a dropdown) and other stuff if possible (: 
Great find!

**Comment by ChristopherKlay** [comment]
**Update:**
Decided to not actually implement this as a search parameter only, but as a filter option for all lists across the entire Stadia store.
All store list pages will feature options to filter by type with `1.8.4` similar to this [demo](https://streamable.com/7fmd84). I did switch the filter and sort position for the final release however, so the filter options are on the left of the sort dropdown. (:

**Reply by superm1991** [reply]
Looks great :)
Thanks a lot!
